### ISSUE TITLE
复现结果差距很大，不知道问题出在哪里，望您能指点一二，非常感谢

### ISSUE THREAD

**Issue by ZMZJ13** [MicroResearchLab/AMP-potency-prediction-EvoGradient#7]
作者您好，感谢您的回复，这是我对attention模型的回归任务的训练代码，
其余三个模型的效果都是正常的，但这个模型在进行EvoGradient.py时的结果和您提供的结果差距很大， 一直没能解决， 不知道问题出在哪里，望您能指点一二，非常感谢
import torch
import numpy as np
from torch import nn
import math
from torch.utils.data import DataLoader
from torch.utils.data import Dataset
import pandas as pd
class PositionalEncoding(nn.Module):
    def __init__(self, len, d_model=20, dropout=0):
        super(PositionalEncoding, self).__init__()
        self.dropout = nn.Dropout(p=dropout)

        pe = torch.zeros(len, d_model)
        position = torch.arange(0, len, dtype=torch.float).unsqueeze(1)
        div_term = torch.exp(torch.arange(0, d_model, 2).float() * (-math.log(10000.0) / d_model))
        pe[:, 0::2] = torch.sin(position * div_term)
        pe[:, 1::2] = torch.cos(position * div_term)
        # pe = pe.unsqueeze(0).transpose(0, 1)
        pe = pe.unsqueeze(0)
        self.register_buffer("pe", pe)

    def forward(self, x):
        x = x + self.pe
        # x = x + self.pe[:,:x.size(0), :]
        return x

class AttentionNetwork(nn.Module):

    def __init__(self, batch_size=128, embedding_size=20, num_tokens=100, num_classes=1, num_heads=4):
        super(AttentionNetwork, self).__init__()
        self.pe = PositionalEncoding(len=num_tokens, d_model=embedding_size)
        self.batch_size = batch_size
        self.embedding_size = embedding_size
        self.num_tokens = num_tokens
        self.num_classes = num_classes
        self.num_heads = num_heads
        self.hidden1 = 20
        self.hidden2 = 60
        self.hidden3 = 20
        self.dropout = 0.2
        self.relu = nn.ReLU()

        self.LN = nn.LayerNorm(normalized_shape=self.hidden1)
        self.fc1 = nn.Linear(self.embedding_size, self.hidden1)

        self.multihead_att = nn.MultiheadAttention(embed_dim=self.hidden1, num_heads=self.num_heads, batch_first=1, dropout=self.dropout)
        self.flatten = nn.Flatten()
        self.fc2 = nn.Linear(self.hidden1 * self.num_tokens, self.hidden2)
        self.fc3 = nn.Linear(self.hidden2, self.hidden3)
        self.new_fc4 = nn.Linear(self.hidden3, self.num_classes)

        self.dropout = nn.Dropout(self.dropout)

    def forward(self, x, mask):
        x = self.pe(x)
        x = self.fc1(x)

        mask = mask.to(torch.bool)
        x, w1 = self.multihead_att.forward(x, x, x, key_padding_mask=mask)

        x = self.flatten(x)
        x = self.fc2(x)
        x = self.dropout(x)

        x = self.relu(x)
        x = self.fc3(x)
        x = self.relu(x)
        x = self.dropout(x)
        x = self.new_fc4(x)

        return x
trainPath = "./data/regression/train.csv"

mydict = {"A": 0, "C": 1, "D": 2, "E": 3, "F": 4, "G": 5, "H": 6, "I": 7, "K": 8, "L": 9, "M": 10, "N": 11, "P": 12, "Q": 13, "R": 14, "S": 15, "T": 16, "V": 17, "W": 18, "Y": 19}
myInvDict = dict([val, key] for key, val in mydict.items())
def dataProcessPipeline(seq):

    # this function first transform peptide sequences into numerical sequence,
    # transformer it into onehot vector and padding them into a fix length
    # returning the padding vector and mask

    testest = seq
    num_seq = [mydict[character.upper()] for character in seq]

    seq = np.array(num_seq, dtype=int)
    len = seq.shape[0]
    torch_seq = torch.tensor(seq,dtype=torch.long)
    if torch.sum(torch_seq[torch_seq < 0]) != 0:
        print(torch_seq[torch_seq < 0])
        print("wrong seq:", seq)
        print(testest)
    onehotSeq = torch.nn.functional.one_hot(torch_seq, num_classes=20)
    # onehotSeq = torch.nn.functional.one_hot(c
    pad = torch.nn.ZeroPad2d(padding=(0, 0, 0, 100 - len))
    mask = np.zeros(100, dtype=int)
    mask[len:] = 1
    mask = torch.tensor(mask)

    pad_seq = pad(onehotSeq)

    return pad_seq, mask
class TrainDataset(Dataset):
    def __init__(self, data_path, transform=dataProcessPipeline):
        df = pd.read_csv(data_path, header=0)

        self.df = df

        self.seqs = list(self.df["Sequence"])
        self.values = self.df["value"]

        self.values.loc[self.values > MAX_MIC] = MAX_MIC
        self.values = list(self.values)
        self.transform = transform

    def __getitem__(self, idex):
        seq = self.seqs[idex]
        num_seq, mask = self.transform(seq)
        label = self.values[idex]

        return num_seq, mask, label

    def __len__(self):
        return len(self.seqs)
    

if __name__ == '__main__':
    batch_size = 128
    MAX_MIC = math.log10(8192)
    trainData = TrainDataset(data_path=trainPath)
    train_loader = DataLoader(dataset=trainData, batch_size=batch_size, shuffle=True, num_workers=4)
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    model = AttentionNetwork(batch_size=128, embedding_size=20, num_tokens=100, num_classes=1, num_heads=4).to(device)
    criterion = nn.MSELoss()
    optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
    # 训练过程
    num_epochs = 500

    for epoch in range(num_epochs):
        model.train()  # 切换到训练模式
        total_loss = 0.0

        for num_seq, mask, labels in train_loader:
            num_seq = num_seq.float().to(device)  # 移动到设备
            mask = mask.to(device)  # 使用 mask 而不是 masks
            labels = labels.float().to(device)
            optimizer.zero_grad()  # 清除梯度
            # 前向传播
            outputs = model(num_seq,mask)
            loss = criterion(outputs, labels.unsqueeze(1)) # 计算损失
            loss.backward()  # 反向传播
            optimizer.step()  # 更新权重

            total_loss += loss.item()

        average_loss = total_loss / len(train_loader)
        print(f'Epoch [{epoch + 1}/{num_epochs}], Loss: {average_loss:.4f}')
        torch.save(model,r'D:\Astudy\AMP-1\data\Attention_regression.pth')
    print("Training complete!")
